### PR TITLE
clean up hoisting

### PR DESCRIPTION
**Pull request by max-hoffman**
Variety of refactors related to hoisting subqueries and unnesting scopes.

- remove `SingleSelectRel`
- empty join expansion
- limits + filters should be tested in memo now
- fix to anti join correctness for empty right-hand rel

### CODE REVIEW

File: enginetest/join_op_tests.go [hunk 1]
```diff
@@ -106,6 +106,102 @@ var joinCostTests = []struct {
...
+				Query: "select * from mytable where exists (select * from mytable where i = 1);",
+				Expected: []sql.Row{
+					{1, "first row"},
+					{2, "second row"},
+					{3, "third row"},
+				},
+			},
+			// queries that test subquery hoisting
+			{
+				// case 1: condition uses columns from both sides
```
Comment (fulghum) [inline enginetest/join_op_tests.go:119]: Thanks for adding in these comments to explain the intent of the test. That is super super helpful for understanding what the tests are covering.

File: enginetest/join_op_tests.go [hunk 2]
```diff
@@ -106,6 +106,102 @@ var joinCostTests = []struct {
...
+					{2, 2},
+					{3, 1},
+				},
+			},
+			{
+				// case 4aN: NOT EXISTS condition uses no columns from either side, and condition is true
+				Query:    "/*case4aN*/ select * from ab where not exists (select * from xy where 1 = 1)",
+				Expected: []sql.Row{},
+			},
+			// TODO: uncomment after https://github.com/dolthub/dolt/issues/5522 is fixed
```
Comment (fulghum) [inline enginetest/join_op_tests.go:183]: Do we still need these TODOs? Looks like that issue is fixed and we already uncommented the tests?

**Review by fulghum — APPROVED** [review]
Nice changes. Cool to get rid of the `SelectSingleRel` type. Just a couple minor comments/questions.

File: engine.go [hunk 3]
```diff
@@ -409,7 +409,7 @@ func (e *Engine) analyzeQuery(ctx *sql.Context, query string, parsed sql.Node, b
 				return nil, err
 			}
 			for binding := range bindings {
-				if !usedBindings[binding] {
+				if !usedBindings[binding] && !plan.HasEmptyTable(analyzed) {
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply engine.go:412]: yeah this potentially carries pitfalls, my justification is that assuming we should only convert something into an empty table for good reasons. If the value of a bindvar changes whether something is an empty scope, the conditions for triggering EmptyTable are maybe buggy.

File: enginetest/join_op_tests.go (hunk 1 above)
Comment (max-hoffman) [reply enginetest/join_op_tests.go:119]: @PavelSafronov deserves credit for this and the tests :)

File: sql/analyzer/memo.go [hunk 4]
```diff
@@ -667,6 +665,8 @@ type indexScan struct {
 	parent *joinBase
 }
 
+//go:generate go run ../../optgen/cmd/optgen/main.go -out memo.og.go -pkg analyzer memo memo.go
+
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply sql/analyzer/memo.go:674]: yeah i can put back, i was annoyed scrolling down to change the def and up to retrigger. Whole code needs to be refactored really, there's no reason optgen should take a dependency on analyzer to read a list of strings.

File: sql/analyzer/analyzer.go [hunk 5]
```diff
@@ -451,6 +451,8 @@ func newInsertSourceSelector(sel RuleSelector) RuleSelector {
 // Analyze applies the transformation rules to the node given. In the case of an error, the last successfully
 // transformed node is returned along with the error.
 func (a *Analyzer) Analyze(ctx *sql.Context, n sql.Node, scope *Scope) (sql.Node, error) {
+	//a.Debug = true
```
Comment (PavelSafronov) [inline sql/analyzer/analyzer.go:454]: Remove?

File: sql/analyzer/memo.go [hunk 6]
```diff
@@ -464,16 +464,21 @@ func (e *exprGroup) updateBest(n relExpr, grpCost float64) {
...
 }
 
-func (e *exprGroup) finalize(node sql.Node) sql.Node {
+func (e *exprGroup) finalize(node sql.Node, input sql.Schema) (sql.Node, error) {
 	props := e.relProps
 	var result = node
 	if props.filter != nil {
-		result = plan.NewFilter(props.filter, result)
+		sch := append(input, node.Schema()...)
+		filter, _, err := FixFieldIndexes(e.m.scope, nil, sch, props.filter)
```
Comment (PavelSafronov) [inline sql/analyzer/memo.go:472]: The fact that we need to fix up indexes suggests that maybe the rules are out of order. If we run this rule before the column indexes are applied, we could skip on this update. Is that possible?

**Pull request merged**